File: skellig-junit-runner/src/main/java/org/skellig/runner/junit/report/model/TestStepReportDetails.java
package org.skellig.runner.junit.report.model;

import org.apache.commons.lang3.StringUtils;
import org.skellig.teststep.processing.model.ExpectedResult;
import org.skellig.teststep.processing.model.TestStep;

import java.beans.IntrospectionException;
import java.beans.Introspector;
import java.beans.PropertyDescriptor;
import java.lang.reflect.Field;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class TestStepReportDetails {

    private String name;
    private Object originalTestStep;
    private Object result;
    private String errorLog;

    private TestStepReportDetails(String name, Object originalTestStep, Object result, String errorLog) {
        this.name = name;
        this.originalTestStep = originalTestStep;
        this.result = result;
        this.errorLog = errorLog;
    }

    public String getName() {
        return name;
    }

    public String getErrorLog() {
        return errorLog;
    }

    public boolean isPassed() {
        return errorLog == null || errorLog.equals("");
    }

    public boolean isIgnored() {
        return originalTestStep == null && result == null;
    }

    public String getTestData() {
        String result = "";
        if (originalTestStep instanceof TestStep) {
            Object testData = ((TestStep) originalTestStep).getTestData();
            result = testData != null ? String.valueOf(testData) : "";
        }
        return result;
    }

    public String getValidationDetails() {
        StringBuilder stringBuilder = new StringBuilder();
        if (originalTestStep instanceof TestStep) {
            ((TestStep) originalTestStep).getValidationDetails()
                    .ifPresent(validationDetails -> constructValidationDetails(validationDetails.getExpectedResult(), stringBuilder));
        }
        return stringBuilder.toString();
    }

    public String getProperties() {
        if (originalTestStep != null && !originalTestStep.getClass().equals(TestStep.class) &&
                originalTestStep instanceof TestStep) {
            Map<String, Optional<Method>> properties = getPropertyGettersOfTestStep(originalTestStep.getClass());
            return properties.entrySet().stream()
                    .filter(entry -> entry.getValue().isPresent())
                    .map(this::getPropertyWithValue)
                    .collect(Collectors.joining("\n"));
        }
        return "";
    }

    public String getResult() {
        return result != null ? String.valueOf(result) : "";
    }

    private void constructValidationDetails(ExpectedResult expectedResult, StringBuilder stringBuilder) {
        if (StringUtils.isNotEmpty(expectedResult.getProperty())) {
            stringBuilder.append(expectedResult.getProperty()).append(": ");
        }
        if (expectedResult.getExpectedResult() instanceof List) {
            stringBuilder.append(getValidationType(expectedResult)).append("[\n");
            expectedResult.<List<ExpectedResult>>getExpectedResult()
                    .forEach(item -> constructValidationDetails(item, stringBuilder));
            stringBuilder.append("]\n");
        } else if (expectedResult.getExpectedResult() != null) {
            stringBuilder
                    .append(expectedResult.getExpectedResult().toString())
                    .append("\n");
        }
    }

    private String getValidationType(ExpectedResult expectedResult) {
        return expectedResult.getValidationType().name().toLowerCase();
    }

    private String getPropertyWithValue(Map.Entry<String, Optional<Method>> propertyGetterPair) {
        try {
            Object getterResponse = propertyGetterPair.getValue().get().invoke(originalTestStep);
            if (getterResponse instanceof Optional) {
                getterResponse = ((Optional<?>) getterResponse).orElse(null);
            }
            return getterResponse != null ? propertyGetterPair.getKey() + ": " + getterResponse : "";
        } catch (IllegalAccessException | InvocationTargetException e) {
            return e.getMessage();
        }
    }

    private Map<String, Optional<Method>> getPropertyGettersOfTestStep(Class beanClass) {
        return Stream.of(beanClass.getDeclaredFields())
                .map(Field::getName)
                .collect(Collectors.toMap(fieldName -> fieldName,
                        fieldName -> {
                            try {
                                return Stream.of(Introspector.getBeanInfo(beanClass).getPropertyDescriptors())
                                        .filter(pd -> pd.getReadMethod() != null && fieldName.equals(pd.getName()))
                                        .map(PropertyDescriptor::getReadMethod)
                                        .findFirst();
                            } catch (IntrospectionException e) {
                                return Optional.empty();
                            }
                        }));
    }

    public static class Builder {
        private String name;
        private Object originalTestStep;
        private Object result;
        private String errorLog;

        public Builder withName(String name) {
            this.name = name;
            return this;
        }

        public Builder withOriginalTestStep(Object originalTestStep) {
            this.originalTestStep = originalTestStep;
            return this;
        }

        public Builder withResult(Object result) {
            this.result = result;
            return this;
        }

        public Builder withErrorLog(String errorLog) {
            this.errorLog = errorLog;
            return this;
        }

        public TestStepReportDetails build() {
            return new TestStepReportDetails(name, originalTestStep, result, errorLog);
        }
    }
}

File: skellig-junit-runner/src/main/kotlin/org/skellig/runner/junit/report/model/TestStepReportDetails.kt
package org.skellig.runner.junit.report.model

import org.apache.commons.lang3.time.DurationFormatUtils
import org.skellig.teststep.processing.model.DefaultTestStep
import org.skellig.teststep.processing.model.TaskTestStep
import org.skellig.teststep.processing.model.TestStep
import org.skellig.teststep.processing.util.PropertyFormatUtils

/**
 * Represents the details of a test step report.
 *
 * @param T The type of the original test step.
 * @property name The name of the test step.
 * @property parameters The parameters of the test step.
 * @property originalTestStep The original test step object.
 * @property result The result of the test step.
 * @property errorLog The error log of the test step.
 * @property logRecords The log records of the test step.
 * @property duration The duration of the test step execution in milliseconds.
 */
open class TestStepReportDetails<T>(
    val name: String,
    val parameters: Map<String, Any?>?,
    val originalTestStep: T?,
    val result: Any?,
    val errorLog: String?,
    val logRecords: List<String>?,
    var duration: Long
) {

    fun isPassed(): Boolean {
        return errorLog == null || errorLog == ""
    }

    fun isIgnored(): Boolean {
        return originalTestStep == null && result == null
    }

    open fun getTestData(): String = ""

    open fun getValidationDetails(): String = ""

    fun getDurationFormatted(): String {
        return getFormattedDuration(duration)
    }

    fun getParameters(): String {
        return parameters?.map { "${it.key} = ${it.value}" }?.joinToString("\n") ?: ""
    }

    fun getProperties(): String {
        if (originalTestStep != null && originalTestStep.javaClass != TestStep::class.java && originalTestStep is TestStep) {
            return originalTestStep.toString()
        }
        return ""
    }

    fun getResult(): String {
        return PropertyFormatUtils.toString(result, 0)
    }

    class Builder {
        private var name: String? = null
        private var originalTestStep: Any? = null
        private var result: Any? = null
        private var errorLog: String? = null
        private var logRecords: List<String>? = null
        private var duration: Long = 0
        private var parameters: Map<String, Any?>? = null

        fun withName(name: String?) = apply {
            this.name = name
        }

        fun withOriginalTestStep(originalTestStep: Any?) = apply {
            this.originalTestStep = originalTestStep
        }

        fun withResult(result: Any?) = apply {
            this.result = result
        }

        fun withErrorLog(errorLog: String?) = apply {
            this.errorLog = errorLog
        }

        fun withLogRecords(logRecords: List<String>?) = apply {
            this.logRecords = logRecords
        }

        fun withDuration(duration: Long) = apply {
            this.duration = duration
        }

        fun withParameters(parameters: Map<String, Any?>?) = apply {
            this.parameters = parameters
        }

        fun build(): TestStepReportDetails<*> {
            return when (originalTestStep) {
                is DefaultTestStep -> DefaultTestStepReportDetails(name!!, parameters, originalTestStep as DefaultTestStep, result, errorLog, logRecords, duration)
                else -> TestStepReportDetails(name!!, parameters, originalTestStep, result, errorLog, logRecords, duration)
            }
        }
    }
}

class DefaultTestStepReportDetails(
    name: String,
    parameters: Map<String, Any?>?,
    originalTestStep: DefaultTestStep?,
    result: Any?,
    errorLog: String?,
    logRecords: List<String>?,
    duration: Long
) : TestStepReportDetails<DefaultTestStep>(name, parameters, originalTestStep, result, errorLog, logRecords, duration) {

    override fun getTestData(): String {
        return PropertyFormatUtils.toString(originalTestStep?.testData ?: "", 0)
    }

    override fun getValidationDetails(): String {
        return originalTestStep?.validationDetails?.toString() ?: ""
    }

}

fun getFormattedDuration(duration: Long): String {
    return if (duration > 60000) {
        DurationFormatUtils.formatDuration(duration, "mm.ss.SSS") + " min."
    } else if (duration >= 1000) {
        DurationFormatUtils.formatDuration(duration, "ss.SSS") + " sec."
    } else DurationFormatUtils.formatDuration(duration, "SSS") + " ms."
}